packages/babel-types/src/builders/validateNode.ts

```diff
@@ -1,12 +1,34 @@
-import validate from "../validators/validate";
+import { validateInternal } from "../validators/validate";
 import type * as t from "..";
-import { BUILDER_KEYS } from "..";
+import { BUILDER_KEYS, NODE_FIELDS } from "..";
+import type { FieldDefinitions } from "../definitions/utils";
 
 export default function validateNode<N extends t.Node>(node: N) {
+  if (node == null || typeof node !== "object") return;
+  const fields = NODE_FIELDS[node.type];
+  if (!fields) return;
+
   // todo: because keys not in BUILDER_KEYS are not validated - this actually allows invalid nodes in some cases
   const keys = BUILDER_KEYS[node.type] as (keyof N & string)[];
   for (const key of keys) {
-    validate(node, key, node[key]);
+    const field = fields[key];
+    if (field != null) validateInternal(field, node, key, node[key]);
+  }
+  return node;
+}
+
+export function validateNodeInternal<N extends t.Node>(
+  node: N,
+  info: [FieldDefinitions, string[]],
+) {
+  if (node == null || typeof node !== "object") return;
+
+  const fields = info[0],
+    builderKeys = info[1] as (keyof N & string)[];
+
+  for (const key of builderKeys) {
+    const field = fields[key];
+    if (field != null) validateInternal(field, node, key, node[key]);
```

Comment:
Does the validation for required field still work with this `field != null` check?

Reply:
It will work, I just followed here to check ahead.
https://github.com/babel/babel/pull/15843/files#diff-04a0d87b83569ab9a9d94ee801346cf5443ecf057bf35f4836d4a259e882c8fcR39
But thanks to you, I found another problem, `validateChild` was accidentally skipped. :)